lib/core/include/irods/irods_at_scope_exit.hpp

```diff
@@ -1,31 +1,106 @@
 #ifndef IRODS_AT_SCOPE_EXIT_HPP
 #define IRODS_AT_SCOPE_EXIT_HPP
 
+/// \file
+
 #include <utility>
 
 namespace irods
 {
+    /// A convenient RAII-based class that defers code execution to the exiting of a scope.
+    ///
+    /// This class guards against all possibilities of throwing from the destructor making it
+    /// safe to use in all cases. For cases where the code is known to never throw, prefer
+    /// irods::at_scope_exit_unsafe.
```

Comment:
Would it be worth mentioning the performance bump here as well, like in the commit message?

Reply:
I think I should remove any mention of performance here because I haven't verified it.

I'll plan to remove that statement from the commit message.

Reply:
Done